### PR TITLE
Added Binary Search Tree [Go]

### LINKED ISSUE #527: Binary Search Tree [Go]
#### This issue is about [Binary Search Tree](https://en.wikipedia.org/wiki/Binary_search_tree).

- [x] Issue Title is in form `Algo/DS Name [Language]`.
- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.
- [x] This Algo/DS is worth adding to this repository.
- [x] Assign this issue to me (I want to work on this).
	- [x] I ensure that I am not already assigned to 2 or more issues.
- [x] I will send my PR only after I'm assigned to this issue by any maintainer.
- [x] I understand that I will be unassigned from this issue if I fail to send a PR for this within a week.

### PR DESCRIPTION
**Pull request by praffn**
Fixes #527 

#### By submitting this pull request I confirm I've read and complied with the below declarations.

- [x] I have read the [Contribution guidelines](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md) and I am confident that my PR reflects them.
- [x] I have followed the [coding guidelines](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md#cs) for this project.
- [x] My code follows the [skeleton code structure](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md#sample).
- [x] This pull request has a descriptive title. For example, `Added {Algorithm/DS name} [{Language}]`, not `Update README.md` or `Added new code`.
- [x] This pull request will be closed if I fail to update it even once in a continuous time span of 7 days.

### CODE REVIEW

**Comment by singhpratyush** [comment]
Hi! Thanks for showing interest in contributing to this repository.

Please make sure that you have ticked the points mentioned in PR description correctly.

Thanks again!

**Comment by praffn** [comment]
I've added it to the readme

File: binary_search_tree/binary_search_tree.go [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+	}
+
+	// Delet a key
+	bst.Delete(12)
+	// Search for same key again
+	_, ok = bst.Search(12)
+	if ok {
+		fmt.Println("This shouldn't happen, since key should have been deleted")
+	} else {
+		fmt.Println("12 was not found :)")
```
Comment (mohitkyadav) [inline binary_search_tree/binary_search_tree.go]: Why it's "12 was not found". Use a variable if you want to use it more than once in a single code. Avoid hardcoding.

File: binary_search_tree/binary_search_tree.go [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+	})
+
+	// Searching for a key. Should return value and whether key was found or not
+	value, ok := bst.Search(12)
+	if ok {
+		fmt.Printf("Key 12 was found! It's value is %d\n", value)
+	}
+
+	// Delet a key
+	bst.Delete(12)
```
Comment (mohitkyadav) [inline binary_search_tree/binary_search_tree.go]: Please implement error handling for this function, it falls into a run time error if passed non existing key.

File: binary_search_tree/binary_search_tree.go [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+	// Print in order. Should print from lowest key to highest
+	bst.Traverse(func(key, value int) {
+		fmt.Printf("Key: %d, Value: %d\n", key, value)
+	})
+
+	// Searching for a key. Should return value and whether key was found or not
+	value, ok := bst.Search(12)
+	if ok {
+		fmt.Printf("Key 12 was found! It's value is %d\n", value)
```
Comment (mohitkyadav) [inline binary_search_tree/binary_search_tree.go]: Don't hardcode 12.

**Pull request merged**